noir-projects/noir-contracts/contracts/app/orderbook_contract/src/main.nr

```diff
@@ -0,0 +1,198 @@
+/// - Since public functions are internal-only, **external contracts cannot access them**, ensuring no external
+///   contract can trigger a reentrant call. This eliminates the following attack vector:
+///   `Orderbook.private_fn --> Orderbook.public_fn --> ExternalContract.fn --> Orderbook.public_fn`.
+#[aztec]
+pub contract Orderbook {
+    use crate::{config::Config, order::Order};
+    use aztec::{
+        event::event_interface::EventInterface,
+        macros::{
+            events::event,
+            functions::{initializer, internal, private, public, utility},
+            storage::storage,
+        },
+        oracle::notes::check_nullifier_exists,
+        prelude::{AztecAddress, Map, PublicImmutable},
+        protocol_types::traits::{FromField, Serialize, ToField},
+        unencrypted_logs::unencrypted_event_emission::encode_event,
+    };
+
+    use token::Token;
+    use uint_note::uint_note::PartialUintNote;
+
+    // The event contains only the `order_id` as the order itself can be retrieved via the `get_order` function.
+    #[derive(Serialize)]
+    #[event]
+    struct OrderCreated {
+        order_id: Field,
+    }
+
+    #[derive(Serialize)]
+    #[event]
+    struct OrderFulfilled {
+        order_id: Field,
+    }
+
+    #[storage]
+    struct Storage<Context> {
+        config: PublicImmutable<Config, Context>,
+        orders: Map<Field, PublicImmutable<Order, Context>, Context>,
+    }
+
+    #[public]
+    #[initializer]
+    fn constructor(token0: AztecAddress, token1: AztecAddress) {
+        storage.config.initialize(Config::new(token0, token1));
+    }
+
+    /// Privately creates a new order in the orderbook
+    /// The maker specifies the tokens and amounts they want to trade
+    #[private]
+    fn create_order(
+        bid_token: AztecAddress,
+        ask_token: AztecAddress,
+        bid_amount: u128,
+        ask_amount: u128,
+        nonce: Field,
+    ) -> Field {
+        let config = storage.config.read();
+
+        // Create the order (this validates the input tokens and amounts).
+        let order = Order::new(config, bid_amount, ask_amount, bid_token, ask_token);
+
+        let maker = context.msg_sender();
+
+        // Transfer tokens from maker to the public balance of this contract.
+        Token::at(bid_token)
+            .transfer_to_public(maker, context.this_address(), bid_amount, nonce)
+            .call(&mut context);
+
+        // Prepare a partial note that will get completed once the order is fulfilled. Note that only the Orderbook
+        // contract can complete the partial note.
+        let maker_partial_note =
+            Token::at(ask_token).prepare_private_balance_increase(maker, maker).call(&mut context);
+
+        // We use the partial note's commitment as the order ID. Because partial notes emit a nullifier with their
+        // commitment when created they are unique, and so this guarantees that our order IDs are also unique without
```

Comment:
i'm not entirely sure this is true / accurate, but it doesn't matter:
- partial notes don't emit a nullifier
- this contract does emit a nullifier per `order_id` (== `partial_note_commitment`)
- a user could theoretically have initialized a commitment in the Token and repeat it here
- this action should have no weird behaviour (other than in the user's PXE to handle the exception)

Reply:
Hello 👋 
partial notes will emit nullifier upon creation once [this PR](https://github.com/AztecProtocol/aztec-packages/pull/14432) is merged.